service/external_paysession.go

```diff
@@ -4,43 +4,80 @@ import (
 	"encoding/json"
 	"fmt"
 	"net/http"
-
-	"github.com/companieshouse/payments.api.ch.gov.uk/models"
+	"regexp"
+	"strconv"
+	"strings"
 
 	"github.com/companieshouse/chs.go/log"
+	"github.com/companieshouse/payments.api.ch.gov.uk/models"
 )
 
 // CreateExternalPaymentJourney creates an external payment session with a Payment Provider that is given, e.g: GovPay
-func CreateExternalPaymentJourney(w http.ResponseWriter, req *http.Request) {
-	if req.Body == nil {
-		log.ErrorR(req, fmt.Errorf("request body empty"))
+func (service *PaymentService) CreateExternalPaymentJourney(w http.ResponseWriter, req *http.Request) {
+	id := req.URL.Query().Get(":payment_id")
+	if id == "" {
+		log.ErrorR(req, fmt.Errorf("payment id not supplied"))
 		w.WriteHeader(http.StatusBadRequest)
 		return
 	}
 
-	requestDecoder := json.NewDecoder(req.Body)
-	var incomingExternalPaymentJourneyRequest models.IncomingExternalPaymentJourneyRequest
-	err := requestDecoder.Decode(&incomingExternalPaymentJourneyRequest)
+	paymentSession, httpStatus, err := (*PaymentService).getPaymentSession(service, id)
 	if err != nil {
-		log.ErrorR(req, fmt.Errorf("request body invalid: [%v]", err))
-		w.WriteHeader(http.StatusBadRequest)
+		w.WriteHeader(httpStatus)
+		log.ErrorR(req, err)
 		return
 	}
 
-	if incomingExternalPaymentJourneyRequest.PaymentMethod != "GovPay" {
-		log.ErrorR(req, fmt.Errorf("payment method not recognised: %v", incomingExternalPaymentJourneyRequest.PaymentMethod))
-		w.WriteHeader(http.StatusBadRequest)
+	paymentJourney := &models.ExternalPaymentJourney{}
+
+	if paymentSession.PaymentMethod == "GovPay" {
+		paymentJourney.NextURL, err = returnNextURLGovPay(paymentSession, id, &service.Config)
+		if err != nil {
+			log.ErrorR(req, fmt.Errorf("error communicating with GovPay: [%s]", err))
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		if paymentJourney.NextURL == "" {
+			log.ErrorR(req, fmt.Errorf("no NextURL returned from GovPay"))
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		err = json.NewEncoder(w).Encode(paymentJourney)
+		if err != nil {
+			log.ErrorR(req, fmt.Errorf("error writing response: %s", err))
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusOK)
 		return
 	}
 
-	paymentJourney := &models.ExternalPaymentJourney{}
-	//TODO: Return next_url from GovPay, hardcoded at the moment
-	paymentJourney.NextURL = "http://gov.uk/paymentjourney"
-	w.Header().Set("Content-Type", "application/json")
-	w.WriteHeader(http.StatusOK)
-	err = json.NewEncoder(w).Encode(paymentJourney)
+	log.ErrorR(req, fmt.Errorf("payment method, [%s], for resource [%s] not recognised", paymentSession.PaymentMethod, id))
+	w.WriteHeader(http.StatusBadRequest)
+	return
+
+}
+
+func convertToPenceFromDecimal(decimalPayment string) (int, error) {
```

Comment:
As discussed:
* refactor `getTotalAmount` to return totalAmount.StringFixed(2)
* remove the regex checking from this function
* refactor the conversion to pence